lib/Conversion/HandshakeToHW/HandshakeToHW.cpp

```diff
@@ -272,225 +317,422 @@ static Operation *checkSubModuleOp(mlir::ModuleOp parentModule,
-  auto ports = getPortInfoForOp(rewriter, oldOp);
-  // todo: HWModuleOp; for this initial commit we'll leave this as an extern op.
-  return rewriter.create<HWModuleExternOp>(
-      parentModule.getLoc(), rewriter.getStringAttr(getSubModuleName(oldOp)),
-      ports);
-}
+namespace {
 
-//===----------------------------------------------------------------------===//
-// HW Top-module Related Functions
-//===----------------------------------------------------------------------===//
+// Input handshakes contain a resolved valid and (optional )data signal, and
+// a to-be-assigned ready signal.
+struct InputHandshake {
+  Value valid;
+  std::shared_ptr<Backedge> ready;
+  Value data;
+};
 
-static hw::HWModuleOp createTopModuleOp(handshake::FuncOp funcOp,
-                                        ConversionPatternRewriter &rewriter) {
-  llvm::SmallVector<PortInfo, 8> ports = getPortInfoForOp(
-      rewriter, funcOp, funcOp.getArgumentTypes(), funcOp.getResultTypes());
+// Output handshakes contain a resolved ready, and to-be-assigned valid and
+// (optional) data signals.
+struct OutputHandshake {
+  std::shared_ptr<Backedge> valid;
+  Value ready;
+  std::shared_ptr<Backedge> data;
+};
 
-  // Create a HW module.
-  auto hwModuleOp = rewriter.create<hw::HWModuleOp>(
-      funcOp.getLoc(), rewriter.getStringAttr(funcOp.getName()), ports);
+struct UnwrappedIO {
+  llvm::SmallVector<InputHandshake> inputs;
+  llvm::SmallVector<OutputHandshake> outputs;
+
+#define IOAccessor(Type, container, member, name)                              \
+  llvm::SmallVector<Type> name() {                                             \
+    llvm::SmallVector<Type> result;                                            \
+    for (auto &container : container) {                                        \
+      result.push_back(container.member);                                      \
+    }                                                                          \
+    return result;                                                             \
+  }
+  // Accessors for getting all the valids, readys, and datas of an IO set.
+  IOAccessor(Value, inputs, valid, getInputValids);
+  IOAccessor(std::shared_ptr<Backedge>, inputs, ready, getInputReadys);
+  IOAccessor(Value, inputs, data, getInputDatas);
+  IOAccessor(std::shared_ptr<Backedge>, outputs, valid, getOutputValids);
+  IOAccessor(Value, outputs, ready, getOutputReadys);
+  IOAccessor(std::shared_ptr<Backedge>, outputs, data, getOutputDatas);
+};
 
-  // Remove the default created hw_output operation.
-  auto outputOps = hwModuleOp.getOps<hw::OutputOp>();
-  assert(std::distance(outputOps.begin(), outputOps.end()) == 1 &&
-         "Expected exactly 1 default created hw_output operation");
-  rewriter.eraseOp(*outputOps.begin());
+// A class containing a bunch of syntactic sugar to reduce builder function
+// verbosity.
+// @todo: should be moved to support.
+struct RTLBuilder {
```

Comment:
Thanks for going down this route! I think this is a good step forward along the lines of what we discussed in the ODM.